Added test for extracting zip files with invalid characters in Windows

Danyy427 opened the pull request:
Relates to #67201

Commit:
Added test for extracting zip files with invalid characters in Windows

Commit:
Removed faulty test case

danmoseley commented:
Thanks. We will need the fix as well before we can merge this 😸

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -103,6 +103,22 @@ public void Windows_ZipWithInvalidFileNames_ThrowsArgumentException(string zipNa
+        [Theory]
+        [InlineData("InvalidWindowsFileNameChars.zip", "Test______________________________________.txt")]
+        [InlineData("InvalidWindowsFileNameChars.zip", "Test______________________________________/TestText1______________________________________.txt")]
+        [InlineData("InvalidWindowsFileNameChars.zip", "Test/normalText.txt")]
```
this will run the test 3 times. you can just run the test once (no InlineData) and check all three files are found.

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -103,6 +103,22 @@ public void Windows_ZipWithInvalidFileNames_ThrowsArgumentException(string zipNa
+        [InlineData("InvalidWindowsFileNameChars.zip", "Test______________________________________/TestText1______________________________________.txt")]
+        [InlineData("InvalidWindowsFileNameChars.zip", "Test/normalText.txt")]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void Windows_ZipWithInvalidFileNames(string zipFile, string testPath)
```
it would be helpful to include a comment here showing what the file paths are inside the zip. maybe just what I pasted into the runtime-assets PR you did.

Commit:
Removed redundant test cases, Added contents of the InvalidWindowsFileNameChars.zip

Danyy427 commented:
@danmoseley I removed all of the ```[InlineData]``` and changed the test to a ```[Fact]``` and tested for all three of the files by hardcoding them. I hopefully didn't make a mistake because I wrote the code directly on GitHub. Once the tests are complete and fail, I will start working on the sanitization.

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -103,6 +103,30 @@ public void Windows_ZipWithInvalidFileNames_ThrowsArgumentException(string zipNa
+        {
+            var testDirectory = GetTestFilePath();
+            ZipFile.ExtractToDirectory(compat("InvalidWindowsFileNameChars.zip"), testDirectory);
+            Assert.True(File.Exists(testDirectory + "Test______________________________________.txt"));
```
Please use `Path.Combine` not `+` - see other tests.

danmoseley commented:
Test looks good with comments above. Please also push up the fix now, that makes this test pass locally for you.

Commit:
Replaced "+"s to Path.Combine for forming the test file path.

danmoseley commented:
BTW, it's fine doing things this way for your first PR 😄 but generally we would push up a PR when it's all done -- meaning tests+fix. We verify that the tests fail without the fix, but we do it locally beforehand.

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -103,6 +103,30 @@ public void Windows_ZipWithInvalidFileNames_ThrowsArgumentException(string zipNa
+        /// This test checks whether or not ZipFile.ExtractToDirectory() is capable of handling filenames
+		/// which contain invalid path characters in Windows.
+        ///  Archive:  InvalidWindowsFileNameChars.zip
+        ///  Length      Date    Time    Name
+        /// ---------  ---------- -----   ----
+        ///        0  2022-03-27 13:44   Test/
+        ///       12  2022-03-27 13:44   Test/normalText.txt
+        ///        0  2022-03-27 13:44   Test"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_/
+        ///       14  2022-03-27 13:44   Test"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_/TestText1"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_.txt
+        ///        0  2022-03-27 13:44   TestEmpty/
+        ///       14  2022-03-27 13:44   TestText"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_.txt
```
```suggestion
        ///  Test/
        ///  Test/normalText.txt
        ///  Test"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_/
        ///  Test"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_/TestText1"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_.txt
        ///  TestEmpty/
        ///  TestText"<>|^A^B^C^D^E^F^G^H^I^J^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z^[^\^]^^^_.txt
```

Commit:
Update src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs

Danyy427 commented:
@danmoseley Oh, I thought when you said to write the tests first and verify it fails, I thought I was supposed to verify it fails on the GitHub repo rather than my local computer. That was the reason I asked if I should do it one go or separately. 

I was going to start writing the fix but I cannot build the repo now due to some technical issue [technical issue](https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1) apparently. Once that is fixed and I can build the repo I will start the fix and probably push it by tomorrow. 

BTW, thank you for helping, I am quite unfamiliar with all of this.

Danyy427 commented:
@danmoseley It doesn't build
``` 
GET https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1
The remote server returned an error: (421) Misdirected Request.
Retrying. Waiting for 3 seconds before next attempt (2 of 5).
GET https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1
The remote server returned an error: (421) Misdirected Request.
Retrying. Waiting for 7 seconds before next attempt (3 of 5).
GET https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1
The remote server returned an error: (421) Misdirected Request.
Retrying. Waiting for 15 seconds before next attempt (4 of 5).
GET https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1
The remote server returned an error: (421) Misdirected Request.
Retrying. Waiting for 31 seconds before next attempt (5 of 5).
GET https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.ps1
The remote server returned an error: (421) Misdirected Request.
Unable to download file in 5 attempts.
```
The link says the following 

```
<h2>Our services aren't available right now</h2><p>We're working to restore all services as soon as possible. Please check back soon.</p>0wKJEYgAAAACW4qqrJ8M1QK2oV6FN05MfQU1TMDRFREdFMTkxMQBlMWRmMDcwYS1hZTQ0LTRjMGItYTU0Yi1jNDkzODA0ZTRkOWY=
```

Commit:
Added code to use ValidFullName across several files

Commit:
Merge branch 'Danyy427-ZipFile-Test' of https://github.com/Danyy427/runtime into Danyy427-ZipFile-Test

Commit:
Removed weird character

Commit:
Added Sanitization

Commit:
Fixed Typo

Commit:
Fixed wrong if statement

Commit:
Fixed wrong filename in test

Commit:
Added sanitization as a method

The tests added pass with this build

Commit:
Merge branch 'Danyy427-ZipFile-Test' of https://github.com/Danyy427/runtime into Danyy427-ZipFile-Test

Commit:
Fixed tests that didn't pass

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileValidFullName_Unix.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        internal static string SanitizeZipFilePath(string zipPath)
+        {
+            StringBuilder builder = new StringBuilder(zipPath);
+            for (int i = 0; i < zipPath.Length; i++)
+            {
+                if ((int)builder[i] == 0)
+                {
+                    builder[i] = '_';
+                }
+            }
+            return builder.ToString();
```
This is shorter and more efficient (eg., no allocations in happy path)
```suggestion
            return zipPath.Replace('\0', '_')
```

although, do we have any evidence that nulls are ever found in file paths in valid real world zip files? the situation with invalid Windows file characters is quite different, because that occurs when you extract from a valid archive made from a Unix system. But null characters are invalid everywhere.

danmoseley replied:
Eh, I think it's fine to leave the null sanitization as it's a natural extension and it's only a matter of time before someone somehow encounters this.

Danyy427 replied:
The tests included null characters so I didn't leave sanitization out for null characters. Also good for completeness I guess

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Create.cs:
```diff
@@ -428,9 +428,19 @@ public async Task UpdateAddFile()
         public void Windows_ZipWithInvalidFileNames_ThrowsException(string zipName, string paramName)
```
The name of this test method is now wrong. But can we just delete these tests now?

Danyy427 replied:
We don't need to delete it, I will fix the name tho.

danmoseley replied:
It's just duplicating your new tests right?

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -90,30 +93,42 @@ public void Unix_ZipWithOSSpecificFileNames(string zipName, string fileName)
+            
+            var testDirectory = GetTestFilePath();
+            ZipFile.ExtractToDirectory(compat("InvalidWindowsFileNameChars.zip"), testDirectory);
+            Assert.True(File.Exists(Path.Combine(testDirectory, "TestText______________________________________.txt")));
```
would it be cleaner to extract a local method like

```c#
string CheckExists(testDirectory, path)
{
    Assert.True(File.Exists(Path.Combine(testDirectory, path)));
}
```

Danyy427 replied:
Would I be placing this method directly into the test file or into testUtilities?

danmoseley replied:
Within this method as a local function?

https://devblogs.microsoft.com/premier-developer/dissecting-the-local-functions-in-c-7/

Danyy427 replied:
There are other places where this is used. Maybe I can put it in test utils and replace it everywhere. Or I can just use local function? Whatever you prefer

danmoseley replied:
I don't think it's worth replacing it everywhere. In fact, we can leave the code as you're written it if you like. Not important.

danmoseley commented:
Looks good just a couple comments.

Danyy427 commented:
@danmoseley I will wrap it up and commit

Commit:
Cleaned up code

Danyy427 commented:
@danmoseley I added that local function. Changed the wrong method name. Added the sanitization for Unix that I forgot to add to Zipfile.Create and simplified the sanitization on UNIX to that single line thing.

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -90,30 +93,48 @@ public void Unix_ZipWithOSSpecificFileNames(string zipName, string fileName)
+            ZipFile.ExtractToDirectory(compat("WindowsInvalid_FromWindows.zip"), testDirectory);
+            CheckExists(testDirectory, "aa_b_d");
+
+
```
```suggestion
```

danmoseley commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs:
```diff
@@ -70,9 +70,12 @@ public void ExtractOutOfRoot(string entryName)
+        public void Unix_ZipWithInvalidFileNames(string zipName)
         {
-            AssertExtensions.Throws<ArgumentException>("path", () => ZipFile.ExtractToDirectory(compat(zipName) + ".zip", GetTestFilePath()));
+            var testDirectory = GetTestFilePath();
```
I still don't see what this test, and all the tests you modified in Zipfile.Create.cs, are doing that your new test isn't now doing. Why can't we delete all of those existing tests?

Danyy427 replied:
I do not know _what_ they are doing per see and why they were added in the first place but I didn't want to remove them. I just fixed them and left them as is

danmoseley replied:
They were testing that extracting a zip file containing a file with a path with a null character, threw ArgumentException. It no longer does this, and your new test verifies that these cases are sanitized, so the tests have no value. Please remove these 3 tests

Danyy427 replied:
Yep, I have the exact same code in the original test anyway.

Danyy427 replied:
Are there 3 tests? I could only find two. One for windows and one for Unix

danmoseley replied:
Doh, you're right. 

Can this Unix test verify that the sanitization _didn't_ happen (except for the nulls)? Basically copy/paste exactly Windows_ZipWithInvalidFileNames, but change the expected strings.

Danyy427 replied:
We have a test like that I think. `Unix_ZipWithOSSpecificFileNames` in Zipfile.Extract

danmoseley replied:
That only checks `<` and `>`. I don't have a strong opinion. Your call

danmoseley replied:
I'll sign off as-is 😄

Danyy427 replied:
Well, we only sanitize nulls anyway, if the code worked as expected before, meaning we could extract zip files in UNIX with the correct result, the code we changed shouldn't affect that. We only added sanitization for null char in Unix and we test that already so I will say it is fine? I guess?

Commit:
Removed unnecessary tests

Danyy427 commented:
@danmoseley Oh, forgot to mention a very important problem. If you look at `System.IO.Compression.ZipFile.csproj` you can see that I have added windows sanitization to both net7.0-windows and to net7.0. Now I do not know how that will affect things but that seemed slightly off to me. Would this affect anything?
```
  <!-- Windows specific files -->
  <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'windows' or '$(TargetPlatformIdentifier)' == ''">
    <Compile Include="System\IO\Compression\ZipFileValidName_Windows.cs" />
  </ItemGroup>
```

danmoseley commented:
Hmm, @ViktorHofer can you advise? This library previously targeted Unix, Browser, and everything else. Now we needed to add a Windows-specific implementation, and it targets Unix, Browser, Windows, and everything else. What meaning is "everything else" here? Should we take that TFM out?

Danyy427 commented:
@danmoseley Do have a solution? Are we removing net7.0 from TF? I think we should as it would not be safe to assume that a system has windows filename restrictions or at least it is safer to assume that it has no restrictions (but NULL of course).

danmoseley commented:
@Danyy427 I need TFM guidance here from an expert like @ViktorHofer on Monday. The code changes are fine.

Danyy427 commented:
@danmoseley ok, I guess we just wait now :)

Danyy427 commented:
@danmoseley Oh, I just realized something. Shouldn’t we also be sanitizing (or modifying atleast) filenames like `CON` or `NUL`. As far I know they are reserved in windows as well. Perhaps that’s beyond the scope of this issue/pull request though.

danmoseley commented:
>  Shouldn’t we also be sanitizing (or modifying atleast) filenames like CON or NUL

No. I think I mentioned in the issue, Windows valid filename rules are complex and of course it also depends on the file system. Ultimately if you have an archive containing files with those unusual names, you'll get an IOException and you can extract them individually to the names of your choice.

danmoseley commented:
@Danyy427 if you are interested in another contribution after this, you might want to browse these https://github.com/dotnet/runtime/issues?q=is%3Aissue+is%3Aopen+label%3Aeasy
(I added label easy because you're just starting out in this repo 😃

ViktorHofer commented:
> Now we needed to add a Windows-specific implementation, and it targets Unix, Browser, Windows, and everything else. What meaning is "everything else" here? Should we take that TFM out?

There are two options:
- Choose one of those three platform implementations to be the default (the platform agnostic tfm). Previously that default was the Windows build.
- If none of these three platforms is a good default for any new platform that we would eventually introduce that would not apply to Unix, Browser and Windows then you might want to make the platform agnostic tfm a PNSE build.

On the phone right now, will post an example on how to do that tomorrow.

Danyy427 commented:
@danmoseley I would be more than happy to continue contributing after this one.

danmoseley commented:
@ViktorHofer thanks. Why would we introduce a PNSE TFM when browser+windows+unix covers everything?

Simplest thing then seems to leave the Windows build as the platform agnostic one. I think that means
* revert the TargetFrameworks line
* change `<ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'windows' or '$(TargetPlatformIdentifier)' == ''">` to `<ItemGroup Condition="'$(TargetPlatformIdentifier)' == ''">`

Commit:
Change agnostic TFM back to Windows

Commit:
spacing

Danyy427 commented:
@danmoseley Asking just out of curiosity, why do we assume windows to be the default when the windows filename rules are the exception? Wouldn't sanitizing only the null character be better?

danmoseley commented:
> why do we assume windows to be the default when the windows filename rules are the exception?

As far as I'm aware, the only target that doesn't match browser and Unix is Windows, so it doesn't matter. If we add a new target, we would need to change other things too. I am not sure what the implications are of changing the default one, although @ViktorHofer would know.

Danyy427 commented:
@danmoseley I don't know if it would change anything code-wise but I was thinking as a design.

danmoseley commented:
Ok it sounds like we should default to the Unix behavior as you suggested @Danyy427 . Do you want to make that change?

Commit:
Changed default target framework to Unix

ViktorHofer commented on src/libraries/System.IO.Compression.ZipFile/src/System.IO.Compression.ZipFile.csproj:
```diff
@@ -1,7 +1,7 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <TargetFrameworks>$(NetCoreAppCurrent)-Unix;$(NetCoreAppCurrent)-Browser;$(NetCoreAppCurrent)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Browser;$(NetCoreAppCurrent)-Unix</TargetFrameworks>
```
This makes sure that the Unix is picked as the default.

```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Browser;$(NetCoreAppCurrent)</TargetFrameworks>
```

Danyy427 replied:
Shouldn't we also remove '$(TargetPlatformIdentifier)' == 'Unix'` if we make this change

Danyy427 replied:
Ok, thanks pushing the changes.

Commit:
Defaulted framework to unix

ViktorHofer commented on src/libraries/System.IO.Compression.ZipFile/src/System.IO.Compression.ZipFile.csproj:
```diff
@@ -1,7 +1,7 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <TargetFrameworks>$(NetCoreAppCurrent)-Unix;$(NetCoreAppCurrent)-Browser;$(NetCoreAppCurrent)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Browser;$(NetCoreAppCurrent)</TargetFrameworks>
```
@eerhardt as you added the Browser and Unix configurations in https://github.com/dotnet/runtime/commit/5b0c6dd5756eea63a57b1f69b182f7386da0f631, do you remember why you needed a separate Browser build?

Was it just to satisfy the build graph (because at that time there was no rid-less default implementation) or because you actually compile differently for Browser?

If the former, then we want to remove the Browser tfm.

Danyy427 replied:
@ViktorHofer done

Commit:
Removed Browser

ViktorHofer commented on src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Create.cs:
```diff
@@ -288,7 +288,7 @@ public void InvalidDates()
                 fileWithBadDate.Create().Dispose();
                 fileWithBadDate.LastWriteTimeUtc = new DateTime(1970, 1, 1, 1, 1, 1);
-
+                
```
```suggestion

```

ViktorHofer reviewed (approved):
From an infrastructure perspective this looks great. Thanks

Commit:
Update src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Create.cs

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

Danyy427 commented:
@danmoseley Are we supposed to do anything else to get this thing to merge?

danmoseley commented:
@Danyy427 nope, just letting tests run.

@dotnet/dnceng something happened here with Azdo https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/1711221/logs/662

ulisesh commented:
Thanks for reporting the problem @danmoseley. I haven't seen this error before but I'll create an issue to monitor this error.

I see that you already retried the build, hopefully it succeeds this time

danmoseley commented:
Thanks @Danyy427. You ran into one or two more problems (on our side) than most PR's encounter!

Danyy427 commented:
@danmoseley It’s okay I guess. Thanks for all the help you have given, it has been most useful. I will most certainly comtinue contributing. And this will certainly not be the last time I will give you a headache from all the questions. :)

danmoseley commented:
Always happy to help. As I mentioned earlier, from experience I highly recommend new folks get a number of easy, short PR's merged before tackling anything larger. Partly just because it takes time to get used to a repo's style, etc.

Danyy427 commented:
@danmoseley Yeah, probably that’s foe the best. I browsed a couple of those easy and up-for-grabs. There are some which are like a year old and almost forgotten. I may continue with those perhaps as early as tomorrow
